optimizely/event/event_processor.py

```diff
@@ -0,0 +1,211 @@
+from ..logger import NoOpLogger
+
+ABC = abc.ABCMeta('ABC', (object,), {'__slots__': ()})
+
+
+class EventProcessor(ABC):
+  """ Class encapsulating event_processor functionality. Override with your own processor
+  providing process method. """
+
+  @abc.abstractmethod
+  def process(user_event):
+    pass
+
+
+class BatchEventProcessor(EventProcessor, Closeable):
+  """
+  BatchEventProcessor is a batched implementation of the EventProcessor.
+  The BatchEventProcessor maintains a single consumer thread that pulls events off of
+  the blocking queue and buffers them for either a configured batch size or for a
+  maximum duration before the resulting LogEvent is sent to the EventDispatcher.
+  """
+
+  _DEFAULT_QUEUE_CAPACITY = 1000
+  _DEFAULT_BATCH_SIZE = 10
+  _DEFAULT_FLUSH_INTERVAL = timedelta(seconds=30)
+  _DEFAULT_TIMEOUT_INTERVAL = timedelta(seconds=5)
+  _SHUTDOWN_SIGNAL = object()
+  _FLUSH_SIGNAL = object()
+  LOCK = threading.Lock()
+
+  def __init__(self,
+                event_dispatcher,
+                logger,
+                default_start=False,
+                event_queue=None,
+                batch_size=None,
+                flush_interval=None,
+                timeout_interval=None):
+    self.event_dispatcher = event_dispatcher or default_event_dispatcher
+    self.logger = logger or NoOpLogger()
+    self.event_queue = event_queue or queue.Queue(maxsize=self._DEFAULT_QUEUE_CAPACITY)
+    self.batch_size = batch_size if batch_size is not None and batch_size > 0 else self._DEFAULT_BATCH_SIZE
+    self.flush_interval = timedelta(milliseconds=flush_interval) if flush_interval is not None and flush_interval > 0 \
+                            else self._DEFAULT_FLUSH_INTERVAL
+    self.timeout_interval = timedelta(milliseconds=timeout_interval) if timeout_interval is not None and \
+                            timeout_interval > 0 else self._DEFAULT_TIMEOUT_INTERVAL
+    self._disposed = False
+    self._is_started = False
+    self._current_batch = list()
+
+    if default_start is True:
+      self.start()
+
+  @property
+  def is_started(self):
+    return self._is_started
+
+  @property
+  def disposed(self):
+    return self._disposed
+
+  def _get_time_in_ms(self, _time=None):
+    return int(round((_time or time.time()) * 1000))
+
+  def start(self):
+    if self.is_started and not self.disposed:
+      self.logger.warning('Service already started')
+      return
+
+    self.flushing_interval_deadline = self._get_time_in_ms() + self._get_time_in_ms(self.flush_interval.total_seconds())
+    self.executor = threading.Thread(target=self._run)
+    self.executor.setDaemon(True)
+    self.executor.start()
+
+    self._is_started = True
+
+  def _run(self):
+    """ Scheduler method that periodically flushes events queue. """
+    try:
+      while True:
+        if self._get_time_in_ms() > self.flushing_interval_deadline:
+          self.logger.debug('Deadline exceeded; flushing current batch.')
```

Comment:
we can remove this log as said by mike in ruby sdk PR [comment](https://github.com/optimizely/ruby-sdk/pull/191/files/096ae95ca7d8f68cf913b5c9583dccf22063db9d#r314019298)